### ISSUE TITLE
Characteristic not yet found. Skipping..

### ISSUE THREAD

**Issue by v0r73x** [stoner221/homebridge-led-strip-ble#1]
Hi stoner221,

I tried to get your code run but after install and successful discovery of device(s) I am not able to control them.

log of discovered devices:

[2020-12-20 15:33:30] [LED_Strip] Discovered:  undefined  |UUID:  ffff80058273
[2020-12-20 15:33:30] [LED_Strip] Discovered:  undefined  |UUID:  14363031996d
[2020-12-20 15:33:30] [LED_Strip] Discovered:  undefined  |UUID:  767bac5ec71b
[2020-12-20 15:33:30] [LED_Strip] Discovered:  undefined  |UUID:  467348ffed4a
[2020-12-20 15:33:32] [LED_Strip] Discovered:  undefined  |UUID:  4adda20dd4c7

log of issue when I try to change state:
[2020-12-20 15:34:13] [LED_Strip] Read Characteristic not yet found. Skipping..
[2020-12-20 15:34:13] [LED_Strip] Read Characteristic not yet found. Skipping..
[2020-12-20 15:34:13] [LED_Strip] Read Characteristic not yet found. Skipping..
[2020-12-20 15:34:13] [LED_Strip] Characteristic not yet found. Skipping..
[2020-12-20 15:34:13] [LED_Strip] Read Characteristic not yet found. Skipping..

config from homebridge config.json:
{
            "accessory": "LED_Strip",
            "name": "LED_Strip",
            "address": "ffff80058273",
            "read_service_uuid": "ffd0",
            "write_service_uuid": "ffd5",
            "read_char_uuid": "ffd4",
            "write_char_uuid": "ffd9"
        }

I have happylight (mobileapp) used to capture by wireshark the btsnoop.log to find the code syntax I need to use.in your code (index.js) I found also the syntax for on and off in your code and hope (guess) it will be compatible with my lamps as well.

The way I get my lamps controlled is at the moment by putting on and off by this command:

gatttool -b FF:FF:80:05:82:73 --char-write-req --handle=0x0007 --value=cc2333
gatttool -b FF:FF:80:05:82:73 --char-write-req --handle=0x0007 --value=cc2433

I tried also by the IOS app lightblue and are able to on/off the lamp

- discover device in LightBlue
- go into UUID:FFD5 0xFFD9 (write without response)
- write new value 0xCC2333 
- and lamp goes on
- write new value 0xCC2433
- and lamp goes off

is there a way to do a more detailed debug why it couldn't find the Characteristic? I am very appreciated if you can help me to get it working. 

best regards and 4th of advent,

v0r73x

**Comment by stoner221** [comment]
Hey @v0r73x so I added some more logging for the Characteristic try running `homebridge -D` and you should get more info let me know how this goes.

**Comment by v0r73x** [comment]
Hi stoner221,

tried with /usr/local/bin/homebridge -D -U /var/lib/homebridge but not more helpful output here to see:

[2020-12-20 18:55:30] [LED_Strip] Characteristic not yet found. Skipping..
[2020-12-20 18:55:30] [LED_Strip] Set Power State: true
[2020-12-20 18:55:30] [LED_Strip] Starting write to bulb
[2020-12-20 18:55:30] [LED_Strip] Characteristic not yet found. Skipping..
[2020-12-20 18:55:31] [LED_Strip] Set Power State: false
[2020-12-20 18:55:31] [LED_Strip] Starting write to bulb
[2020-12-20 18:55:31] [LED_Strip] Characteristic not yet found. Skipping..
[2020-12-20 18:55:32] [LED_Strip] Set Power State: true
[2020-12-20 18:55:32] [LED_Strip] Starting write to bulb
[2020-12-20 18:55:32] [LED_Strip] Characteristic not yet found. Skipping..
[2020-12-20 18:55:37] [LED_Strip] Set Power State: false
[2020-12-20 18:55:37] [LED_Strip] Starting write to bulb
[2020-12-20 18:55:37] [LED_Strip] Characteristic not yet found. Skipping..


is there a way to test just on and off in the script? I am sure my bulbs using the same protocol / code as yours and it will be great to extend/improve your plugin since triones (as mine is) is very often used in smart bulb products

**Comment by stoner221** [comment]
Hey @v0r73x i didn’t see a message in there about `Found Characteristics: ` can you check your local code on line 226 and see if there is a debug log message there if not you might need to update to the latest release.

**Comment by v0r73x** [comment]
Hi Stoner221,

I updated to version 0.07 and see only this in the log:

![initial_boot](https://user-images.githubusercontent.com/48632087/102749187-d63bd200-4363-11eb-968f-3b3b8b93e97c.png)

**Comment by v0r73x** [comment]
attached the whole log i received 
- at the end you see that I tried to power on/off the bulb in homekit
![Bildschirmfoto vom 2020-12-21 08-12-57](https://user-images.githubusercontent.com/48632087/102749519-71cd4280-4364-11eb-8155-93af8d48f1d2.png)

and the installed hb plugin version:

![Bildschirmfoto vom 2020-12-21 08-14-58](https://user-images.githubusercontent.com/48632087/102749650-accf7600-4364-11eb-9116-b54bce542156.png)


and finally the code  beginning  from line #226
![code226](https://user-images.githubusercontent.com/48632087/102752824-7268d780-436a-11eb-82b2-77454c66d0c2.png)

**Comment by stoner221** [comment]
Hey @v0r73x ill go through later today and put in detailed logging in for the whole process to see where your running into issues. I will respond to the comment when finished.

**Comment by v0r73x** [comment]
ok perfect - btw: I build up an rpi3 as test environment and did a test with actual lts node (14 something) also discover is working but set/write/discover of characteristic fails.

It will be so cool if we get it work :) 

thanks so lot,
Rob3rt

**Comment by v0r73x** [comment]
let me know if i can help here by triggering something to test noble. the install was without issues or errors on the new test rpi3  (debian buster). I also installed homebridge as standalone without ui and installed your plugin also by cli. since i am very interested to get it work I will support as best I can ()_____)______________)~~~

**Comment by stoner221** [comment]
Hey @v0r73x  added more logging and even looked at the issue a bit longer I noticed its not even connecting to your device so I added logging at the start to see what's going on. I updated NPM and GIT with the latest shoot me your debug log when you get a chance.

**Comment by v0r73x** [comment]
Hi Stoner221,

again, many thanks for helping :) 

I installed the 0.08 Version of your plugin and done a debug all.
I received a lot of information but I guess this ones are the ones which interesting for us:

![image](https://user-images.githubusercontent.com/48632087/102857527-dfdb3d80-4428-11eb-877a-ffad06d6ffad.png)


and why I have this bluetooth MAC address on 2 different "addresses" ? That is strange or?
![image](https://user-images.githubusercontent.com/48632087/102858176-1ebdc300-442a-11eb-868f-e80962079678.png)



and that is, what I get from log if i try to power on/off in homekit app 
![image](https://user-images.githubusercontent.com/48632087/102858612-e2d72d80-442a-11eb-9f92-6cf83d5849b2.png)

**Comment by v0r73x** [comment]
i checked again in wireshark and it looks we have the same characteristics so the code must work, or?

this code (i tried in Lightblue to set this string) make the LED light blue

0000   02 03 00 0e 00 0a 00 04 00 12 07 00 56 02 fe ff
0010   00 f0 aa



![image](https://user-images.githubusercontent.com/48632087/102890044-10889a80-445c-11eb-97cb-fbb8f94c1cc9.png)

I guess there is something wrong with my noble installation - but since I tried 3 different rpi's without success I guess I forgot something.

I tried now the node advertisement-discovery.js from noble examples - that works - so in general noble do what it should:

![image](https://user-images.githubusercontent.com/48632087/102902093-c230c700-446e-11eb-8415-a0bce2b64b54.png)


is there a possibility to read out what kind of BT protocols my rpi3 support?
What version of node you have running ? 
I have also a other BT dongle but have no idea how to switch to HCi1 since hci0 is default.

on the rpi it is also a desktop installed. could it be that this causes issues with noble to get access to the bt dongle? 
should i deactivate it ? thanks for your help!

**Comment by stoner221** [comment]
I mean it could and to be honest my node skills are sub par at best and especially with noble but it looks like noble is having issues with the bind so you might try different setups. I have an ubuntu server with a bt dongle attached to a usb nothing too special. Let me know if there is more i can do.

**Comment by v0r73x** [comment]
Hi Stoner221,
I built now my own plugin with full Triones support in Homekit. I used some of your code snippets as idea since I am security network engineer and haven't so deep knowledge of coding. Hope that is ok? Mostly the plugin I "modified" based on https://github.com/lucavb/homebridge-magic-blue-bulb and I sill use it for private purposes. If I have the OK of you and Lukas I will upload this plugin to github to make it usable for all who want to use it. Thanks for the basic idea with noble and best whishes for XMAS and happy new year :)
